assets + stylesheet assets

eoghanmurray opened the pull request:
Medium to large enhancement to rrweb, building upon #1239 

---

## Asset Events

Assets are a new type of event that embody a serialized version of a http resource captured during snapshotting. Some examples are images, media files and stylesheets. Resources can be fetched externally (from cache) in the case of a href, or internally for blob: urls and same-origin stylesheets. Asset events are emitted subsequent to either a FullSnapshot or an IncrementalSnapshot (mutation), and although they may have a later timestamp, during replay they are rebuilt as part of the snapshot that they are associated with.  In the case where e.g. a stylesheet is referenced at the time of a FullSnapshot, but hasn't been downloaded yet, there can be a subsequent mutation event with a later timestamp which, along with the asset event, can recreate the experience of a network-delayed load of the stylesheet.

## Assets to mitigate stylesheet processing cost

In the case of stylesheets, rrweb does some record-time processing in order to serialize the css rules which had a negative effect on the initial page loading times and how quickly the FullSnapshot was taken (see https://pagespeed.web.dev/). These are now taken out of the main thread and processed asynchronously to be emitted (up to `processStylesheetsWithin` ms) later.  There is no corresponding delay on the replay side so long as the stylesheet has been successfully emitted.

## Asset Capture Configuration

The `captureAssets` configuration option allows you to customize the asset capture process. It is an object with the following properties:

[NOTE June 2026: currently simplifying the config]
~- `objectURLs` (default: `true`): This property specifies whether to capture same-origin `blob:` assets using object URLs. Object URLs are created using the `URL.createObjectURL()` method. Setting `objectURLs` to `true` enables the capture of object URLs.~

~- `origins` (default: `false`): This property determines which origins to capture assets from. It can have the following values:~
~- `false` or `[]`: Disables capturing any assets apart from object URLs, stylesheets (unless set to false) and images (if that setting is turned on).~
~- `true`: Captures assets from all origins.~
~- `[origin1, origin2, ...]`: Captures assets only from the specified origins. For example, `origins: ['https://s3.example.com/']` captures all assets from the origin `https://s3.example.com/`.~

- `images` (default: `false` or `true` if `inlineImages` is true in rrweb.record config): When set, this option turns on asset capturing for all images irrespective of their origin. Unless this configuration option is explicitly set to `false`, images may still be captured if their src url matches the `origins` setting above.

- `stylesheets` (default: `'without-fetch'`): When set to `true`, this turns on capturing of all stylesheets and style elements via the asset system irrespective of origin. The default of `'without-fetch'` is designed to match with the previous `inlineStylesheet` behaviour, whereas the `true` value allows capturing of stylesheets which are otherwise inaccessible due to CORS restrictions to be captured via a fetch call, which will normally use the browser cache.  Unless this is explicitly set to `false`, a stylesheet will be captured if it matches via the `origins` config above.

- `stylesheetsRuleThreshold` (default: `0`): only invoke the asset system for stylesheets with more than this number of rules. Defaults to zero (rather than say 100) as it only looks at the 'outer' rules (e.g. could have a single media rule which nests 1000s of sub rules). This default may be increased based on feedback.

- `processStylesheetsWithin` (default: `2000`): This property defines the maximum time in milliseconds that the browser should delay before processing stylesheets. Inline `<style>` elements will be processed within half this value. Lower this value if you wish to improve the odds that short 'bounce' visits will emit the asset before visitor unloads page.  Set to zero or a negative number to process stylesheets synchronously, which can cause poor scores on e.g. https://pagespeed.web.dev/ ("Third-party code blocked the main thread").

Juice10 committed:
Add Asset event type and capture assets

Juice10 committed:
Add test to prove player works

Juice10 committed:
Rename `assetCaptureConfig` to `assetCapture`

Juice10 committed:
Document `assetCapture` config

Juice10 committed:
Create asset event for assets that failed to load

Juice10 committed:
Move types from rrweb-snapshot to @rrweb/types

Juice10 committed:
WIP asset manager for replay

Juice10 committed:
Add placeholder tests

Juice10 committed:
Fix image source attribute in rebuild function

Juice10 committed:
Update path in tsconfig.json

Juice10 committed:
Fix asset loading in Replayer

Juice10 committed:
Add todo tests

Juice10 committed:
Add asset mutation events for replayer test

Juice10 committed:
Add replay for cached asset on attribute changes

Juice10 committed:
Fix syntax error

Juice10 committed:
Move assetCapture params to @rrweb/types and add them to meta events

Juice10 committed:
Add asset capture config to meta events

Juice10 committed:
NodeType was moved to @rrweb/types

Juice10 committed:
Upgrade jest and rollup in rrweb-snapshot

Juice10 committed:
Add check types

Juice10 committed:
Update snapshots to include assetCapture settings

Juice10 committed:
Mark `inlineImages` as deprecated in favor of `assetCapture`

Juice10 committed:
Fix rrdom build

Juice10 committed:
Fix deprecated test

Juice10 committed:
Rename `assetCapture` to `captureAssets`

Also deprecate a couple inlineImages tests

Juice10 committed:
Remove console.log statements from AssetManager
class

Juice10 committed:
Rename assetCapture to captureAssets in rrweb test

Juice10 committed:
Check if url is of cacheable origin

Juice10 committed:
Test: Add captureAssets config to meta event

Juice10 committed:
object urls aren't stable so lets remove them

Juice10 committed:
inlineImages should work next to captureAssets for now

Juice10 committed:
Fix asset caching bug

Juice10 committed:
captureAssets on live-stream by default

Juice10 committed:
Fix asset URLs in events file

Juice10 committed:
Update deprecated inlineImages option to use
captureAssets directly

Juice10 committed:
Deprecate inlineImage config
option.

Juice10 committed:
Update snapshot to reflect current captureAssets

Juice10 committed:
Make test less flaky

Juice10 committed:
Make test less flaky

Juice10 committed:
Cache attribute support for BODY, TABLE,
TD, TR, TH, TBODY, THEAD, image, feImage, and
cursor elements

Juice10 committed:
Remove unused 'inlineImages' option

Juice10 committed:
Refactor AssetManager constructor signature

Juice10 committed:
Remove console.log statements in asset.test.ts

Juice10 committed:
Apply formatting changes

Juice10 committed:
Refactor reset method signature to accept optional
config parameter

Juice10 committed:
Fix deserialized size comparison in deserializeArg
test

Juice10 committed:
Fix asset manager to only load assets from
cacheable origins

Juice10 committed:
Update defaultPort in startServer function

Juice10 committed:
Make test more reliable

Juice10 committed:
Add failing test for src attribute

Juice10 committed:
Remove failed asset event

Juice10 committed:
Add support for asset manager in rrdom
And remove attribute while asset is loading

Juice10 committed:
Change asset manager path

Juice10 committed:
Fix import path for AssetManager in asset-unit.test.ts

Juice10 committed:
Move isAttributeCacheable tests to rrweb-snapshot

Juice10 committed:
Make assetManager optional

Juice10 committed:
upgrade rrdom-nodejs to jest 29

Juice10 committed:
Update Jest and ts-jest versions to 29 in rrvideo

Rebase note: yarn install to resolve a yarn.lock conflict, but it didn't succeed; carrying on

Juice10 committed:
add @types/prettier dependency

Juice10 committed:
Add esModuleInterop and allowSyntheticDefaultImports to tsconfig.json

Juice10 committed:
Add support for updating srcset attribute in
AssetManager

Juice10 committed:
Add support for SVG elements in AssetManager

Juice10 committed:
Drop support for caching xlink:href attributes
xlink:href is deprecated
Most modern browsers rewrite `xlink:href` to `href` which is currently supported
Tested in Firefox, Safari & Chrome

Juice10 committed:
Remove console.log statement in
AssetManager.manageAttribute

Juice10 committed:
Linux and Mac generate slightly different screenshots

Juice10 committed:
Remove syntax error

Juice10 committed:
Add wait for RAF before evaluating snapshots

Juice10 committed:
Changeset: Added support for Asset Event and capturing many
different types of assets (not just img#src)

Juice10 committed:
Update puppeteer version in package.json

Juice10 committed:
Fix cross-origin iframe message forwarding test

Juice10 committed:
Deprecate `inlineImages` and introduce `captureAssets` option

Juice10 committed:
Docs: add examples to `getUrlsFromSrcset`

Juice10 committed:
Revert accidental change to `backToNormal`

Juice10 committed:
Move @rrweb/types to devDependencies for rrweb-snapshot

Juice10 committed:
Remove console.log statements in asset-integration.test.ts

Juice10 committed:
Update packages/rrweb/src/replay/asset-manager/index.ts

Co-authored-by: Eoghan Murray <eoghan@getthere.ie>

eoghanmurray committed:
Delay call until all attributes are present on the new node, in case the AssetManager needs to interrogate other attributes to determine how to manage, e.g. <link with href depending on rel="stylesheet"

eoghanmurray committed:
Rename Cacheable -> Capturable

eoghanmurray committed:
Indicate the presence of deferred assets using new `rr_captured_*` attributes instead of storing config on the Meta event

 - a better approach as discussed with Justin and Yanzhen

eoghanmurray committed:
Revert attributes in non-live mode and ensure this happens when we can determine that Assets won't be arriving later

eoghanmurray committed:
Only record certain iframes as assets as we don't want to override the existing nested iframe mechanism

eoghanmurray committed:
Ensure we correctly distinguish between html <iframes> and other exotic types that inline images or pdfs

eoghanmurray committed:
Bump timeouts here as I was getting failures when running whole test suite concurrently

eoghanmurray committed:
Fixup: we don't need template strings here

eoghanmurray committed:
Add an `asset` type which includes the element; a bit of prep work for handling stylesheet assets

eoghanmurray committed:
This function was replaced by getSourcesFromSrcset which ultimately derives from https://github.com/rrweb-io/rrweb-snapshot/pull/74

eoghanmurray committed:
Create assetStatus type as per Justin's request

eoghanmurray committed:
Refactor and prepare for some changes: Deal with the difference between `srcset` and other url attributes in one place

Juice10 committed:
Fix types as per suggestions from code review

Juice10 committed:
Update packages/rrweb/src/replay/canvas/2d.ts

eoghanmurray committed:
Don't rename attrs of assets which will be refused according to initial recording config, i.e. the origins don't match. Should make the 'refused' status redundant

eoghanmurray committed:
New config option inlineStylesheet='all' in order to fetch stylesheets (that are otherwise blocked due to e.g. CORs issues) as assets

eoghanmurray committed:
Simplify onAssetDetected function signature

eoghanmurray committed:
Don't double capture blobs as _cssText and an asset. Fixes test 'captures stylesheets with `blob:` url'

eoghanmurray committed:
Prefer to inlineStylesheets using Assets so that we can take advantage of requestIdleCallback and not block the main thread during snapshotting

eoghanmurray committed:
No need to search through document.styleSheets as the sheet property is available https://developer.mozilla.org/en-US/docs/Web/API/HTMLLinkElement/sheet

eoghanmurray committed:
Allow certain assets to be captured synchronously during mutation

eoghanmurray committed:
Add a slight delay as per #1455

eoghanmurray committed:
Add setting to control how quickly stylesheet assets get emitted - solves real problem in 'can record style text mutations' where we need to have the asset events emitted so that we can replay correctly

eoghanmurray committed:
Don't queue up asset callbacks twice; was happening after `replayer.pause` in 'can record style text mutations' test

 - actual fix is to skip in `applyEventsSynchronously` as the event has already been handled by preloadAllAssets on FullSnapshot event

eoghanmurray committed:
We can rebuild styles immediately without the async/promise delay that is necessary for images. This fixes the 'can record style text mutations' test where we are scrubbing around the replayer synchronously and expect the DOM to keep up

eoghanmurray committed:
When we rebuild styles immediately from a snapshot, ensure we are correctly using cssTextSplits by applying to the serializedNode rather than the DOM node. Another fix the 'can record style text mutations' test

eoghanmurray committed:
Didn't realize we can use the 'as' keyword to treat an rrdom element as a real element - otherwise rrweb-snapshot would need to import RRStyleElement which would create a cyclic dependency

eoghanmurray committed:
Improvements to comments on buildStyleNode (don't want to rebase as is on another branch)

eoghanmurray committed:
Fix up an issue with the assumption that origins applies only to images

eoghanmurray committed:
Remove tests that are duplicated minus the 'if inlineImages is on' part - they are basically testing the same thing as the iframe versions still have <img> elements. These tests were also previously marked DEPRECATED due to association with older inlineImages setting

eoghanmurray committed:
Don't emit style mutations if we haven't serialized the style sheet yet - the asset event will include the final content anyway

eoghanmurray committed:
Replicate the 'removing style elements' test but with an asset

eoghanmurray committed:
Catch and fix omission of absoluteToStylesheet processing on style text node mutations

eoghanmurray committed:
Add a try/catch here as we may not be reading anchor.href

eoghanmurray committed:
Capture <style> element css via an asset event to avoid main thread processing cost during initial snapshotting

eoghanmurray committed:
Explore how we capture assets that have disappeared from the DOM

eoghanmurray committed:
Add some notes after considering changing these to use assets

eoghanmurray committed:
Add explanation as to why we are patching createObjectURL

eoghanmurray committed:
Idea from Justin: Store cssText as an array rather than with the extra `splits` parameter when it is an asset so that there's a uniform way of processing it

eoghanmurray committed:
Allow usage of requestIdleCallback to be turned off altogether by setting a zero or negative value for `config.processStylesheetsWithin`

eoghanmurray committed:
requestIdleCallback: Process inline <style> elements before <link rel="stylesheet">

eoghanmurray committed:
Simplify by passing the inlineStylesheets config setting around as part of the captureAssets config where possible

eoghanmurray committed:
Restore `inlineImages` as a first class config option which bypasses the `origins` mechanism to attempt to inline all images. The previous method will still be used in standalone snapshots

eoghanmurray committed:
These settings are not 'inlining' in the snapshot but rather producing new asset events

eoghanmurray committed:
Transfer the legacy default setting into a special 'without-fetch' in config.stylesheets, as I want that to match inlineImages in that `true` will attempt to fetch all stylesheets

eoghanmurray committed:
Enable all capturing of stylesheets via assets to be turned off

eoghanmurray committed:
Add a setting to enable turning off of assets for smaller stylesheets (in terms of outer rules)

eoghanmurray committed:
Properly handle a loading CORS stylesheet as a mutation + asset rather than just an asset

eoghanmurray committed:
Rewrite of assets.md and move out of recipes into core docs

eoghanmurray committed:
Apply formatting changes

eoghanmurray committed:
Satisfy eslint

eoghanmurray committed:
Satisfy typescript by creating an Interface to acknowledge the extra synthetic property on the style element

eoghanmurray committed:
Don't just re-assert 'detected' in the explanatory comment

YunFeng0817 committed:
Update assets.zh_CN.md

Thanks for adding this document. I tested that even the translation result of Chatgpt4 is not good enough.
So I changed some words to make the doc more readable for Chinese users.

Juice10 committed:
Apply formatting changes

eoghanmurray committed:
Prefer waitForRAF over timeout as the latter operates in puppeteer rather than browser context

Juice10 committed:
Apply formatting changes

Juice10 committed:
Apply formatting changes

Juice10 committed:
Trigger CI again

eoghanmurray committed:
Fix typo in deprecated test

eoghanmurray committed:
Remove doubly added prettier reference

eoghanmurray committed:
Fixup expected server url; presumably this is a side effect of switch to vite

eoghanmurray committed:
Fixup test: maybe vite doesn't incorrectly produce this dupe?

eoghanmurray committed:
Continue fixes from #1510

eoghanmurray committed:
Extra pause needed after switch from jest to vitest, when running tests in parallel - not needed when running test by itself

eoghanmurray committed:
Fixups to imports; I'm not sure where this all went wrong in the rebase process

eoghanmurray committed:
fixup! Caught a case where a text mutation, which bypasses `_cssText`, wasn't getting rebuilt with `adaptCssForReplay`

eoghanmurray committed:
This was removed in single-style-capture but 'reintroduced' in this history via asset-event branch

eoghanmurray committed:
Fix that the `false` default on `shouldIgnoreAssets` should have been a `true` (ignore by default) - an empty (or maybe partial) captureAssets config where origins was unspecified was producing the error

eoghanmurray committed:
Fix that we would have tried to capture an asset with a malformed url with { origins: true }

eoghanmurray committed:
Honour the 'omits css contents for asset managed stylesheet' test:

    // presence of onAssetDetected means we should get
    // rr_captured_href (with contents promised later - i.e. using rrweb/record)

Not sure when this test started failing

eoghanmurray committed:
Improve consistency in terms of where in the test the asset appears

eoghanmurray committed:
Apply formatting changes

eoghanmurray committed:
Prior to this PR, we weren't previously storing blank _cssText; could be a fixup to 'Capture <style> element css via an asset event...'

eoghanmurray committed:
Improve test reliability

eoghanmurray committed:
Think we need tighter processing for the tests to pass in the CI - presume that as the machines are quicker there, the tests complete before stylesheets are processed

jeffdnguyen commented on packages/rrweb-snapshot/src/snapshot.ts:
```diff
@@ -632,53 +679,100 @@ function serializeElementNode(
+    ) {
+      let cssText = stringifyCssRules(styleRules);
+      if (cssText) {
+        cssText = absoluteToStylesheet(cssText, sheetBaseHref);
```
For a style tag that imports a stylesheet that has a import rule. The href here would be incorrect for any urls imported in the nested stylesheet.

In the example below I would expect `absoluteToStylesheet` to return something like `url(\"https://local.pendo.io:8081/browser-tests/temp-resources/Vorname.otf\")`

but instead I get something like `url(\"https://local.pendo.io:8081/browser-tests/guides/Vorname.ttf\")`
<img width="560" alt="Screenshot 2024-07-11 at 1 43 34 PM" src="https://github.com/rrweb-io/rrweb/assets/26877421/bd23c4d2-9d9d-4c61-8be9-4df99f67ab1d">
<img width="602" alt="Screenshot 2024-07-11 at 1 42 29 PM" src="https://github.com/rrweb-io/rrweb/assets/26877421/53ee04f5-01b8-4325-ae73-0180871f2011">
<img width="709" alt="Screenshot 2024-07-11 at 1 42 58 PM" src="https://github.com/rrweb-io/rrweb/assets/26877421/b5c249a3-55bf-4476-b9ca-8da080c7fa2a">

eoghanmurray replied:
Oh that's a great catch; would it be possible to port over a simplified version of that test?

I can't quite figure out whether this was an issue before (and I haven't come across the code to import nested stylesheets during this PR — if you know where that is, please let me know!)

jeffdnguyen replied:
Digging a little more, I honestly think this is an existing issue. From my understanding when we call `absoluteToStylesheet` we make the assumption that any URLs inside of a style tag must share the same filepath as the document the style tag is on. `sheetBaseHref` in the PR here, `getHref()` on master.

Unless you'd like to handle it here, I would be more than happy to open up a bug to continue digging into this and possibly even take a crack at fixing this.

jeffdnguyen replied:
Ah I apologize, I misspoke. I still think the issue has been around for awhile, but is surfaced here because we started checking all style tags instead of just empty ones. Just wanted to clarify!

<img width="757" alt="Screenshot 2024-07-11 at 2 57 09 PM" src="https://github.com/rrweb-io/rrweb/assets/26877421/2034195d-2cb1-4731-80a1-6916a425fc43">
<img width="720" alt="Screenshot 2024-07-11 at 2 58 40 PM" src="https://github.com/rrweb-io/rrweb/assets/26877421/922c4ace-4b9f-4abd-ba9d-4ff79426059d">

eoghanmurray replied:
We've always been inlining <style> tags, just that previously the serialization happened in `serializeTextNode`, whereas this PR consolidates all the logic up to the `serializeElementNode` level.

OK so I believe this all happens because we can recurse into `rule.styleSheet` when we encounter an `@import` rule ... but we call `absoluteToStylesheet` with a single outer href after the stringification.

I reckon this is a long standing issue, so best to open a new issue, or submit a test case as a new PR independently of this one to see if it fails on current master.  

It'd be easier (for me) to fix it after merging this PR due to the changes to the code (e.g. `stringifyStylesheet` has been renamed to `stringifyCssRules`) however we can do the fix in both branches if this one takes too long to merge.

If I've identified the problem correctly, the fix would be to pass down a current baseHref to stringifyCssRules, so that it does the `url()` rewrite during stringification, and also so that we can change the href as we recurse through the tree.

jeffdnguyen replied:
I think you nailed and that we're on the same page here. Making that change I'm getting the correct expected response. I've opened up a [PR](https://github.com/rrweb-io/rrweb/pull/1533), but still need to create a proper test case.

eoghanmurray replied:
The PR has been merged since so resolving this conversation

eoghanmurray committed:
These tests were showing variation in CI

eoghanmurray committed:
Following https://stackoverflow.com/a/64167731 on rebas

eoghanmurray committed:
Following https://stackoverflow.com/a/64167731 on rebase #2

eoghanmurray committed:
Unknown changes to yarn.lock

eoghanmurray committed:
Apply formatting changes

eoghanmurray committed:
Add a queueing mechanism so that in Live Mode we don't render full snapshots until we receive the stylesheet assets to avoid a flash of unstyled content (fouc)

Juice10 commented on packages/types/src/index.ts:
Note to self, we should pull these type changes into their own PR, to ease the release & maintenance of this PR

eoghanmurray replied:
that was completed in #1593 I believe

eoghanmurray committed:
assetCount would not suffice if multiple FullSnapshots were present in the recording, and the ordering of asset arrival in the replayer was not regular. This change also means we only wait for stylesheets (which have a `timeout` associate with their status), which is good as other assets can be handled asyncrounously by the replayer asset manager without any negative effects (images get a placeholder image, whereas a missing stylesheet affect the entire page rendering)

eoghanmurray committed:
Satisfy typings

eoghanmurray committed:
Apply formatting changes

eoghanmurray committed:
Mark where I think the assetManager resets were supposed to happen as I'm about to remove the reset in favour of keeping assets around

eoghanmurray committed:
More accurate setting of 'allAdded' based on how asset urls compare to that specified in the full snapshot

eoghanmurray committed:
Allow duplicate assets to only be sent once per recording (e.g. in a multi-page session).

 - instead of resetting between FullSnapshots, we instead record assets against a timestamp
 - prefer assets with timestamps subseqent to a snapshot (as if a reset happened)
 - if none can be found, find the most recent prior asset for a url

The asset manager now requires a rudimentary idea of where the replayer is at prior to applying an asset

eoghanmurray committed:
Fixup test: this test doesn't use the stylesheet assets so make sure that we don't delay fullsnapshot rendering awaiting for them

eoghanmurray committed:
Fixup test, there's no significance in asset ordering

Commit:
fix typing issue

Commit:
typed node

eoghanmurray committed:
Store the full URL for <style> assets as from a backend point of view it's important to distinguish between the first style element on one page, and the first style element on another page (which would likely have different content)

eoghanmurray committed:
The `allAdded` flag no longer made sense in the context of multiple fullsnapshots. I don't fully understand the reason for returning the `failed` status; have copied over the commit message (from when allAdded was introduced) as the comment 'we don't expect assets to arrive later'

eoghanmurray committed:
Guard against scenarios where preloadAllAssets is not called (came up in "should be able to modify a node's attribute once asset is loaded" test)

eoghanmurray committed:
Fix for 'should list original url in non-live mode when asset never gets loaded' - I believe this change is needed after the change to always storing the `src` in `rr_captured_src`, but I'm unsure as to why I didn't pick it up earlier; maybe I wasn't running these tests

eoghanmurray committed:
This was failing again, so have reverted

eoghanmurray committed:
Don't capture small stylesheets as assets as timing issues are causing nondeterminism in the test (e.g. in 'should nest record iframe' between the emission of the iframe and emission of the style asset which are both (asynchronously) triggered by the fullsnapshot)

eoghanmurray committed:
Ensure blocking class is respected to prevent processing of assets

eoghanmurray committed:
Add settings for explicit control of whether to record video and audio. Add in a missing case where a <picture> element should be considered an image

eoghanmurray committed:
Strip out base64 from asset events also so that they don't show up in output

eoghanmurray committed:
Allow video/audio/images to be explicitly turned off in order to override ['origins'] list

eoghanmurray committed:
Fix patch import

eoghanmurray committed:
'callback' name was no longer obvious to me

eoghanmurray committed:
Add a test to check whether video blobs don't get captured by default

eoghanmurray committed:
Fix that we were overriding the defaults with the empty object, and flag that the test is now failing

eoghanmurray committed:
Fix camelCase error

eoghanmurray committed:
Refactor to avoid an eslint error and an extra call to parentNode

eoghanmurray committed:
Fix some linting errors

eoghanmurray committed:
Avoid a typescript 'Unsafe member access' error (even though it was an assignment). Don't understand this stuff.

eoghanmurray committed:
Try to fix: error TS7053: Element implicitly has an 'any' type because expression of type 'capturedAssetStatuses' can't be used to index type '{ node: serializedNodeWithId; initialOffset: { left: number; top: number; }; }'.

eoghanmurray committed:
Another attempt to placate eslint; there is no failure on my local computer

eoghanmurray committed:
`.attributes` are no longer populated when an element is blocked after "Ensure blocking class is respected to prevent processing of assets"

eoghanmurray committed:
Lack of requestIdleCallback on iOS was causing unexpected event emission ordering due to synchronous emission of an asset during a full snapshot before the FullSnapshot event could be emitted. At least if they explicitly set processStylesheetsWithin=0 then there is a cause relating to this outcome. A further update might ensure that all Assets emitted during a FullSnapshot take on the timestamp of the FullSnapshot

eoghanmurray committed:
Don't capture stylesheet Assets if they have a media attribute which is currently not in force

eoghanmurray committed:
Set AssetEvent timestamps such that they match the FullSnapshot or IncrementalSnapshot (mutation) that generated them.

The logic is that this is the time they were created, even if capture was delayed at record time.
If sorting by timestamp is performed before replay, then so long as the sort algorithm is stable, this will bring the asset events closer to the snapshot that they are associated with, where they are needed for replay

Note: This is a rework of a subset of the code from unmerged #558 allowing timestamps to be set outside of the emission event

eoghanmurray committed:
Fix bug where delayed <style> elements were not getting built

eoghanmurray committed:
As per #1657

eoghanmurray committed:
This changeset is already covered in 2.0

eoghanmurray committed:
Revert "Trigger CI again"

This reverts commit ad187bb86250bb6411d992c7ab6e00600fc65081.

eoghanmurray committed:
Revert something added by Justin in 'Mark `inlineImages` as deprecated in favor of `assetCapture`' as I don't think it's anything germaine to the assets branch...

Claude:

	The new command:
	yarn run concurrently --success=all -r -m=1
		'yarn workspaces-to-typescript-project-references --check'
		'yarn turbo run check-types --continue'

	...does two things at once: checks that TypeScript project references are in sync with the workspace graph (workspaces-to-typescript-project-references --check), then runs the actual
	 type-check via turbo. The -m=1 means they run sequentially (max concurrency 1), so concurrent here is really just using concurrently for its output handling/exit-code logic. It's
	not replacing turbo — turbo still does the actual type-checking.

eoghanmurray committed:
Some inconsequential rearrangements to reduce diff size with Justin's alternative merge of this PR

eoghanmurray committed:
Revert "Fix rrdom build"

This reverts commit 9f511636357450d3451fe4b6000d27e5378a9a98

This should be upstreamed separately if it is valid and not included as part of Assets PR

eoghanmurray committed:
Rename changeset and ensure it also bumps record/replay packages.  We're aiming for backwards compatibility with rrweb-snapshot, so change that to patch

eoghanmurray committed:
Apply formatting changes

eoghanmurray committed:
First version of modifying adoptedStylesheet events to lean on new Asset events to separate content from the adoption event

eoghanmurray committed:
Record some limitations in cross origin iframes + adoptedStylesheets; to follow up

eoghanmurray committed:
Improve the data model by replacing both the `styleIds` or `styles` structures from the original adoptedStylesheet event with a cleaner single array that refers to Assets instead (which may have been already emitted)

eoghanmurray committed:
Add a 50ms delay to adopted stylesheet capture to check that the website is not in the process of programmatically adding rules; we prefer to capture a final stylesheet instead of an initial empty one + a lot of mutations

eoghanmurray committed:
Fix omission where adoptedStylesheets weren't getting transformed in the same way as <style> and <link> elements

eoghanmurray committed:
Make the changeset more succinct

eoghanmurray committed:
Apply formatting changes

eoghanmurray committed:
Remove the `!this.liveMode` check here as I believe it is redundant;

 - if an asset was recorded as 'failed' at record time, then we should respect that even in live mode
 - the case where we synthetically generate a 'failed' status (relating to expectedAssets) already has a `!this.liveMode` guard

eoghanmurray committed:
Enable reversion of <link> elements upon failure; rather than reset the href (these are replayed as <style>) instead embed an import of the original href

eoghanmurray committed:
Protect against reversion of entire srcset when a non-rendered asset never loaded (speculative)

eoghanmurray committed:
Apply formatting changes
